src/lib/uTop.ml

```diff
@@ -41,10 +41,13 @@ let version = "%%VERSION%%"
 let history = LTerm_history.create []
 
 let unopt = function
-        | Some(v) -> v
-        | None -> "~/"
+    | None -> ""
+    | Some(v) -> v
 
-let history_file_name = ref(Some(Filename.concat (unopt M.cache_dir) ".utop_history"))
+let () =
+    try ignore(Sys.is_directory (unopt M.cache_dir)) with Sys_error(_) -> Sys.mkdir (unopt M.cache_dir) 755
+
+let history_file_name = ref (Option.map (fun cache_dir -> Filename.concat cache_dir "history") M.cache_dir)
```

Comment:
```ml
let () = Option.iter (fun cache_file ->
    let dir = Filename.dirname cache_file in
    Sys.mkdir dir 0o600; (* This isn't equivalent to `mkdir -p` which is what should be implemented *)
  ) !history_file_name
```

[Here](https://github.com/dbuenzli/bos/blob/master/src/bos_os_dir.ml#L17) you have an equivalent of `mkdir -p` in OCaml.